test_examples.sh

```diff
@@ -20,25 +34,86 @@ function cleanup {
 }
 trap cleanup EXIT
 
+export ALL_OPTS=$TEMPDIR/all_opts
+if [[ "$LONG" == "--long" ]]; then
+  p() { [ $# -eq 0 ] && echo || (shift; p "$@") | while read r ; do echo -e "$1,$r\n$r"; done }
+  echo -e $OPTS | p `cat` | sort | uniq | sed 's/,$//' | tail -n +2 > $ALL_OPTS
+else
+  echo "all" > $ALL_OPTS
+fi
+
+
 # Test file $1
 function runtest {
-  echo "running test $1"
-  yes 0 | $SOURIR "$1" --quiet --opt all > $TEMPDIR/$1.opt.out && \
-  yes 0 | $SOURIR "$1" --quiet > $TEMPDIR/$1.out && \
-  diff $TEMPDIR/$1.out $TEMPDIR/$1.opt.out > /dev/null && \
-  yes 1 | $SOURIR "$1" --quiet --opt all > $TEMPDIR/$1.opt.out && \
-  yes 1 | $SOURIR "$1" --quiet > $TEMPDIR/$1.out && \
-  diff $TEMPDIR/$1.out $TEMPDIR/$1.opt.out > /dev/null
-}
+  TEST=$1
+  INPUT=$2
+  OPT=$3
+
+  OUT="$TEMPDIR/$TEST-$INPUT"
+  BASELINE_OUT="$OUT-base.out"
+  OPT_OUT="$OUT-$opt.out"
+
+  # echo "running test $1"
+
+  if [ ! -f $BASELINE_OUT ]; then
+    yes "$INPUT" | $SOURIR "$TEST" --quiet &> "$BASELINE_OUT"
+    if [ $? -ne 0 ]; then
+      echo "Baseline run failed on $TEST with input $INP"
+      echo " ----- LOG ----------------------------------------"
+      cat $BASELINE_OUT
+      echo " --------------------------------------------------"
+      exit 255
+    fi
+  fi
+
+  yes "$INPUT" | $SOURIR "$TEST" --opt $OPT --quiet &> "$OPT_OUT"
+  if [ $? -ne 0 ]; then
+    echo "Opt run failed on $TEST with input $INP and opts $OPT"
+    echo " ----- LOG ----------------------------------------"
+    cat $OPT_OUT
+    echo " --------------------------------------------------"
+    exit 255
+  fi
+
+  diff $BASELINE_OUT $OPT_OUT > /dev/null
 
-# Iterate over examples in directory
-ALL_OK=0
-for f in *.sou; do
-  runtest $f
-  if [[ $? -ne 0 ]]; then
-    echo "Example $f failed!"
-    ALL_OK=127
+  if [ $? -ne 0 ]; then
+    echo "Output differs on $NAME with input $INP and opts $OPT"
+    echo " ----- DIFF ---------------------------------------"
+    diff $BASELINE_OUT $OPT_OUT
+    echo " --------------------------------------------------"
+    exit 255
   fi
-done
+}
+
+function status {
+  name=$1
+  done=`wc -l < $STATUS`
+  echo -ne "\e[0K\r[${done}/${NUM_TESTS}] ${name} "
+}
+
+function runtests {
+  TEST=$0
+  status "$TEST"
+  for i in `echo -e $INPUTS`
+  do
+    while read opt
+    do
+      runtest $TEST $i $opt
+    done
+  done < $ALL_OPTS
+  echo $TEST >> $STATUS
+}
+
+export -f runtests runtest status
+
+export NUM_TESTS=`find . -name '*.sou' | wc -l`
+export STATUS="$TEMPDIR/status"
+touch $STATUS
+
+find . -name '*.sou' | xargs -n 1 -P `ncores` bash -c 'runtests $@'
```

Comment:
Currently running the tests takes 1.7s second on my machine when they work. Garbled output is going to make me more than 1.7s second whenever it fails, and make me deal with a problem I don't want to have to deal with. Please try to use the `parallel` command instead (I think pretty much the same syntax should work; the manpage also has a "Function tester" example that describes this use-case), or maybe just drop the parallelization that we don't really need.